prov/rxm/src/rxm_conn.c

```diff
@@ -786,22 +786,109 @@ int rxm_cmap_alloc(struct rxm_ep *rxm_ep, struct rxm_cmap_attr *attr)
 	return ret;
 }
 
+static ssize_t rxm_send_credits(struct fid_ep *ep, size_t credits)
+{
+	struct rxm_conn *rxm_conn =
+		container_of(ep->fid.context, struct rxm_conn, handle);
+	struct rxm_ep *rxm_ep = rxm_conn->handle.cmap->ep;
+	struct rxm_deferred_tx_entry *def_tx_entry;
+	struct rxm_tx_base_buf *tx_buf;
+	struct iovec iov;
+	struct fi_msg msg;
+	ssize_t ret;
+
+	tx_buf = (struct rxm_tx_base_buf *) rxm_tx_buf_alloc(
+		rxm_ep, RXM_BUF_POOL_TX_CREDIT);
+	if (!tx_buf) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
+			"Ran out of buffers from TX credit buffer pool.\n");
+		return -FI_ENOMEM;
+	}
+
+	rxm_ep_format_tx_buf_pkt(rxm_conn, 0, rxm_ctrl_credit, 0, 0, FI_SEND,
+				 &tx_buf->pkt);
+	tx_buf->pkt.ctrl_hdr.type = rxm_ctrl_credit;
+	tx_buf->pkt.ctrl_hdr.msg_id = ofi_buf_index(tx_buf);
+	tx_buf->pkt.ctrl_hdr.ctrl_data = credits;
+
+	if (rxm_conn->handle.state != RXM_CMAP_CONNECTED)
+		goto defer;
+
+	iov.iov_base = &tx_buf->pkt;
+	iov.iov_len = sizeof(struct rxm_pkt);
+	msg.msg_iov = &iov;
+	msg.iov_count = 1;
+	msg.context = tx_buf;
+	msg.desc = &tx_buf->hdr.desc;
+
+	ret = fi_sendmsg(ep, &msg, FI_PRIORITY);
+	if (ret) {
+defer:
+		def_tx_entry = rxm_ep_alloc_deferred_tx_entry(
+			rxm_ep, rxm_conn, RXM_DEFERRED_TX_CREDIT_SEND);
+		if (!def_tx_entry) {
+			FI_WARN(&rxm_prov, FI_LOG_CQ,
+				"unable to allocate TX entry for deferred CREDIT mxg\n");
+			ofi_buf_free(tx_buf);
+			return -FI_ENOMEM;
+		}
+
+		def_tx_entry->credit_msg.tx_buf = tx_buf;
+		rxm_ep_enqueue_deferred_tx_queue(def_tx_entry);
+		return FI_SUCCESS;
+	}
```

Comment:
nit: I would move the defer code outside of the if statement as:
```
    if (!ret)
        return FI_SUCCESS;

defer:
    def_tx_entry = ...
```
It 'undents' most of the code and avoids having a goto jump inside an if statement.